frontend/javascripts/oxalis/view/right-border-tabs/segments_tab/segments_view.tsx

```diff
@@ -582,28 +585,44 @@ class SegmentsView extends React.Component<Props, State> {
       };
     }
     if (prevState.prevProps?.meshes !== meshes) {
-      // Derive the areSegmentsInGroupVisible state so that we know per group
+      // Derive the groupsSegmentsVisibilityStateMap state so that we know per group
       // if it contains only visible elements. This is used to know whether "Show meshes" or
       // "Hide meshes" context item should be shown.
       // If any segment is invisible, set the visibility of the group to false, so that the preferred
       // action is to make all meshes visible.
-      const newVisibleMap: { [groupId: number]: boolean } = {};
-      const segmentGroupsWithRoot = [...segmentGroups, rootGroup];
-      segmentGroupsWithRoot.forEach((group) => {
+      const newVisibleMap: State["groupsSegmentsVisibilityStateMap"] = {};
+
+      const fillNewGroupsSegmentsVisibilityStateMap = (group: SegmentGroup) => {
+        group.children.forEach(fillNewGroupsSegmentsVisibilityStateMap);
+
+        const visibilityEntry = {
+          areSomeSegmentsVisible: group.children.some(
+            (childGroup) => newVisibleMap[childGroup.groupId].areSomeSegmentsVisible,
+          ),
+          areSomeSegmentsInvisible: group.children.some(
+            (childGroup) => newVisibleMap[childGroup.groupId].areSomeSegmentsInvisible,
+          ),
+        };
+
         const segmentsOfGroup = groupToSegmentsMap[group.groupId];
         if (segmentsOfGroup == null) return;
-        const isSomeSegmentLoadedAndInvisible = segmentsOfGroup.some((segment) => {
+        segmentsOfGroup.forEach((segment) => {
```

Comment:
couldn't you still use `some` to avoid inspecting all segments of a group once both bools are true? within the arrow function you could return `return visibilityEntry.areSomeSegmentsVisible && visibilityEntry.areSomeSegmentsInvisible` if I'm not mistaken.

Reply:
Oh right, sounds smart :tada:. On it :+1: